### PR TITLE
Improve cli boot performance by prefering autoload over requires

### PR DESCRIPTION
**Pull request by mwrock**
On my windows system, this improves `chef-client --version` performance by ~2.5x. It will also improve other scenarios to a lesser degree. I think we could yield even more but this is pretty much all the low hanging fruit and then some.

(https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: lib/chef/mixin/openssl_helper.rb [hunk 1]
```diff
@@ -20,7 +20,7 @@ module Mixin
     # various helpers for use with openssl. Currently used by the openssl_* resources
     module OpenSSLHelper
       def self.included(_base)
-        require "openssl" unless defined?(::OpenSSL)
+        autoload :OpenSSL, "openssl"
```
Comment (lamont-granquist) [inline lib/chef/mixin/openssl_helper.rb:23]: you can probably now delete this handler and move it to the top level?

File: lib/chef/util/diff.rb [hunk 2]
```diff
@@ -86,6 +83,9 @@ def diff(old_file, new_file)
       # produces a unified-output-format diff with 3 lines of context
       # ChefFS uses udiff() directly
       def udiff(old_file, new_file)
+        require "diff/lcs"
+        require "diff/lcs/hunk"
```
Comment (lamont-granquist) [inline lib/chef/util/diff.rb:87]: that could probably autoload both on Diff::LCS and don't worry about calling out "hunk" specifically

File: chef-config/lib/chef-config/config.rb [hunk 3]
```diff
@@ -29,7 +29,7 @@
 require_relative "path_helper"
 require_relative "mixin/fuzzy_hostname_matcher"
 
-Mixlib.autoload :Shellout, "mixlib/shellout"
+Mixlib.autoload :ShellOut, "mixlib/shellout"
```
Comment (lamont-granquist) [inline chef-config/lib/chef-config/config.rb:32]: Oh right, nested class constants...  There's gotta be a better syntax to use here...

Comment (lamont-granquist) [reply chef-config/lib/chef-config/config.rb:32]: (it appears there isn't)

Comment (mwrock) [reply chef-config/lib/chef-config/config.rb:32]: I've been poking around and it looks like a better convention is:
```
module Mixlib
    autoload :ShellOut, "mixlib/shellout"
end
```
It srill feels clumsy but it is more precise for the cases where the module` may not be loaded.

Comment (lamont-granquist) [reply chef-config/lib/chef-config/config.rb:32]: yeah that would be my concern, and yeah that would be my suggestion.

(also means you have to know if its a module or a class at the top level, which is mildly annoying)

Comment (lamont-granquist) [reply chef-config/lib/chef-config/config.rb:32]: (although that may only matter for `class Chef` and everything else should be properly a module I hope)

File: lib/chef/util/diff.rb (hunk 2 above)
Comment (mwrock) [reply lib/chef/util/diff.rb:87]: As far as I can tell you can only autoload one require per symbol.

**Comment by phiggins** [comment]
The buildkite CI job should pass if you rebase this on master.

File: lib/chef/util/diff.rb (hunk 2 above)
Comment (lamont-granquist) [reply lib/chef/util/diff.rb:87]: well that's annoying.

i wonder if we need to write a better autoload implementation or look and see what activesupport has that we might be able to copy....

**Review by lamont-granquist — APPROVED** [review]

**Comment by tas50** [comment]
Unrelated failures in unit. Nice stuff @mwrock

**Pull request merged**